### PR TITLE
Api/chat

### PR DESCRIPTION
**Pull request by DipanwitaDey20**
resolved PR

### CODE REVIEW

File: app/Helpers/helpers.php [hunk 1]
```diff
@@ -221,7 +221,9 @@ function milestone_completion_notification($task_id){
 
 function project_paused_notification($project_id){
     $project = Project::where('id', $project_id)->first();
-    $estimate = Estimate::where('project_id', $project_id)->first();
+    $estimate = Estimate::where('project_id', $project_id)
+                        ->where('status', 'project_started')
```
Comment (SankalanSaha) [inline app/Helpers/helpers.php:225]: When the estimate is accepted by the customer then I don't think the estimate status is changed to project_started

Comment (SankalanSaha) [reply app/Helpers/helpers.php:225]: Line 341: Same changes to be made as in this one

File: app/Http/Controllers/Api/ProjectController.php [hunk 2]
```diff
@@ -44,88 +59,152 @@ public function add_project(Request $request){
         try{
             $result = Project::create([
                 'user_id'=> $request->user()->id,
-                'project_address_id' => $request->project_address_id,
+                'builder_category_id' => null,
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:62]: Why are you making the builder_category_id as a required field in validation and then putting null here?

File: app/Http/Controllers/Api/ProjectController.php [hunk 3]
```diff
@@ -29,13 +34,23 @@ public function show(Request $request, Project $project){
 
     public function add_project(Request $request){
         $validator = Validator::make($request->all(), [
-            'project_address_id' => 'required|string',
+            'user_id'=> 'required|bigint',
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:37]: I don't think there is a "bigint" validation rule in laravel

File: app/Http/Controllers/Api/ProjectController.php [hunk 4]
```diff
@@ -44,88 +59,152 @@ public function add_project(Request $request){
...
+                'tradeperson_note' => null,
+                'internal_note' => null,
+                'status' => $request->status,
+                'reviewer_id' => null,
+                'reviewer_status' => null,
+                'reviewer_status_updated_at' => null,
                 'notes'=> $request->notes
             ]);
             if (!$result) {
                 throw ValidationException::withMessages(['message' => 'Something went wrong, please try again!'], 400);
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:86]: Change the status code from 400 to 500

File: app/Http/Controllers/Api/ProjectController.php [hunk 5]
```diff
@@ -44,88 +59,152 @@ public function add_project(Request $request){
...
             }
+            foreach ($request->images as $file) {
+                $fileName = $file->getClientOriginalName();
+                $extension = $file->getClientOriginalExtension();
+                $s3FileName = Str::uuid().'.'.$extension;
+                $file_type = explode('/', mime_content_type($file->getRealPath()))[0];
+                Storage::disk('s3')->put('Testfolder/'.$s3FileName, file_get_contents($file->getRealPath()));
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:93]: Asked You previously to get the folder name from const file

Comment (SankalanSaha) [reply app/Http/Controllers/Api/ProjectController.php:93]: Changes to be made for both line 93 and 94

File: app/Http/Controllers/Api/ProjectController.php [hunk 6]
```diff
@@ -44,88 +59,152 @@ public function add_project(Request $request){
...
+                $projectfile_entry = Projectfile::create([
+                    'project_id'=> $result->id,
+                    'file_type' => $file_type,
+                    'filename' => $s3FileName,
+                    'file_original_name' => $fileName,
+                    'file_extension' => $extension,
+                    'url' => $path
+                ]);
+            }
+
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:105]: Where are the DB::beginTransaction() and DB::commit() statements that I previously asked to add?

File: app/Http/Controllers/Api/ProjectController.php [hunk 7]
```diff
@@ -29,13 +34,23 @@ public function show(Request $request, Project $project){
...
             'contact_mobile_no' => 'required|string',
+            'contact_home_phone' => 'required|string',
             'contact_email' => 'required|string',
+            'postcode' => 'required|string',
+            'county' => 'required|string',
+            'town' => 'required|string',
+            'categories' => 'required|string',
+            'subcategories' => 'required|string',
+            'status' => 'required|string',
+            'images' => 'required|array',
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:52]: Have you checked in the web that if the images is a required field or not?

File: app/Http/Controllers/Api/ProjectController.php [hunk 8]
```diff
@@ -44,88 +59,152 @@ public function add_project(Request $request){
...
+                $extension = $file->getClientOriginalExtension();
+                $s3FileName = Str::uuid().'.'.$extension;
+                $file_type = explode('/', mime_content_type($file->getRealPath()))[0];
+                Storage::disk('s3')->put('Testfolder/'.$s3FileName, file_get_contents($file->getRealPath()));
+                $path = Storage::disk('s3')->url('Testfolder/'.$s3FileName);
+
+                $projectfile_entry = Projectfile::create([
+                    'project_id'=> $result->id,
+                    'file_type' => $file_type,
+                    'filename' => $s3FileName,
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:99]: Check in web if the filename is storing s3filename or the original filename.

File: app/Http/Controllers/Api/ProjectController.php [hunk 9]
```diff
@@ -44,88 +59,152 @@ public function add_project(Request $request){
...
+            $project->subcategories = $request->subcategories;
+            $project->notes = $request->notes;
             if(!$project->save()){
                 return response()->json(['message'=>'Unexpected error, please try after sometimes'],400);
             }
+            //delete images
+            $previousFiles = Projectfile::where('project_id', $project->id)->get();
+
+            foreach ($previousFiles as $previousFile) {
+                $s3Path = 'Testfolder/' . $previousFile->filename;
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:147]: Fetch folder name from const file

Comment (SankalanSaha) [reply app/Http/Controllers/Api/ProjectController.php:147]: Also you need to send the s3 file name here.

File: app/Http/Controllers/Api/ProjectController.php [hunk 10]
```diff
@@ -44,88 +59,152 @@ public function add_project(Request $request){
...
+                Storage::disk('s3')->delete($s3Path);
+                $previousFile->delete();
+            }
+            //update images
+            foreach ($request->images as $file) {
+                $fileName = $file->getClientOriginalName();
+                $extension = $file->getClientOriginalExtension();
+                $s3FileName = Str::uuid().'.'.$extension;
+                $file_type = explode('/', mime_content_type($file->getRealPath()))[0];
+                Storage::disk('s3')->put('Testfolder/'.$s3FileName, file_get_contents($file->getRealPath()));
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:157]: Folder Name from config

File: app/Http/Controllers/Api/ProjectController.php [hunk 11]
```diff
@@ -44,88 +59,152 @@ public function add_project(Request $request){
...
     }
 
-    public function get_file(Request $request){
+    public function cancel_project(Request $request){
         $validator = Validator::make($request->all(), [
-            'project_id' => 'required|Project',
+            'project_id' => 'required',
+            'status' => 'required'
         ]);
-        if ($validator->fails()) {
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:94]: Why are you deleting these lines?

File: app/Http/Controllers/Api/ProjectController.php [hunk 12]
```diff
@@ -44,88 +59,152 @@ public function add_project(Request $request){
...
+    public function paused_project(Request $request){
         $validator = Validator::make($request->all(), [
-            'project_id' => 'required|Project',
-            'file_name' => 'required',
-            'file_extention' => 'required',
-            'url' => 'required',
+            'project_id' => 'required',
+            'status' => 'required'
         ]);
-        if ($validator->fails()) {
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:115]: Why are you deleting these lines?

File: app/Http/Controllers/Api/ProjectController.php [hunk 13]
```diff
@@ -44,88 +59,152 @@ public function add_project(Request $request){
...
             return response()->json($e->getMessage(), 500);
         }
     }
 
-    public function get_file(Request $request){
+    public function cancel_project(Request $request){
         $validator = Validator::make($request->all(), [
-            'project_id' => 'required|Project',
+            'project_id' => 'required',
+            'status' => 'required'
```
Comment (SankalanSaha) [inline app/Http/Controllers/Api/ProjectController.php:181]: Status won't be send from the app if you are creating two APIs for cancelling and pausing project

**Review by SankalanSaha — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Api/ProjectController.php (hunk 2 above)
Comment (DipanwitaDey20) [reply app/Http/Controllers/Api/ProjectController.php:62]: resolved

File: app/Http/Controllers/Api/ProjectController.php (hunk 3 above)
Comment (DipanwitaDey20) [reply app/Http/Controllers/Api/ProjectController.php:37]: resolved

File: app/Http/Controllers/Api/ProjectController.php (hunk 4 above)
Comment (DipanwitaDey20) [reply app/Http/Controllers/Api/ProjectController.php:86]: resolved

File: app/Http/Controllers/Api/ProjectController.php (hunk 9 above)
Comment (DipanwitaDey20) [reply app/Http/Controllers/Api/ProjectController.php:147]: resolved

File: app/Http/Controllers/Api/ProjectController.php (hunk 10 above)
Comment (DipanwitaDey20) [reply app/Http/Controllers/Api/ProjectController.php:157]: resolved

File: app/Http/Controllers/Api/ProjectController.php (hunk 5 above)
Comment (DipanwitaDey20) [reply app/Http/Controllers/Api/ProjectController.php:93]: resolved

File: app/Http/Controllers/Api/ProjectController.php (hunk 13 above)
Comment (DipanwitaDey20) [reply app/Http/Controllers/Api/ProjectController.php:181]: resolved

File: app/Helpers/helpers.php (hunk 1 above)
Comment (DipanwitaDey20) [reply app/Helpers/helpers.php:225]: resolved

File: app/Http/Controllers/Api/ProjectController.php (hunk 6 above)
Comment (SankalanSaha) [reply app/Http/Controllers/Api/ProjectController.php:113]: Resolved.

**Pull request merged**